client/src/App.tsx

```diff
@@ -25,19 +25,29 @@ import { Payload } from './lib/payload';
 import { SearchIcon } from '@chakra-ui/icons';
 import { search } from './lib/search';
 
+function debounce(f, delay) {
+  let lastTimeout = null;
+  return () => {
+    if (lastTimeout) clearTimeout(lastTimeout);
+    lastTimeout = setTimeout(() => {
+      f();
+    }, delay);
+  };
+}
+
 const App: React.ElementType = () => {
   const [value, setValue] = useState<string>('');
   const [payload, setPayload] = useState<Payload | undefined>(undefined);
   const [error, setError] = useState<string | undefined>(undefined);
 
-  const handleInputChange = async (value: string) => {
+  const handleInputChange = debounce(async (value: string) => {
     try {
       setPayload(await search(value));
       setValue(value);
     } catch (error) {
       setError(error instanceof Error ? error.message : 'Unknown Error');
     }
-  };
+  }, 150);
```

Comment:
Not sure if we need this much of a delay, maybe like 10-20ms?

Reply:
Typing at 600cpm there is noticeable delay. 600cpm is about 100ms between each keystroke, so I added a bit

Reply:
Oh I see, misunderstood at first how the debouncing worked, in that case 150ms or higher seems fine. One thing to note is that we don't wanna be setting the input in the debounce, since this adds a delay before the character appears on the screen.

Reply:
One way to do this would be to add a useEffect for `value` that uses the debounce function instead of using the onChange handler.

Reply:
Good point. I think it would be good to make the input uncontrolled by not setting its value at all. Then it will display the correct text and only re-render when necessary.

Reply:
Hmm not sure what you mean by this, we would need to keep track of the input to know it has changed and therefore able to be used in a `useEffect`. I was thinking something like this:

```js
useEffect(() => {
  if (value) {
    const handler = setTimeout(async () => {
      try {
        setPayload(await search(value));
      } catch (error) {
        setError(error instanceof Error ? error.message : 'Unknown Error');
      }
    }, 100);
    return () => clearTimeout(handler);
  }
}, [value]);
```

We'd then get rid of the current input `onChange` handler, and replace everything with calls to `setValue`.

Reply:
I just committed a change that makes the input uncontrolled so you can see what I mean. Also I just found the hook [useDeferredValue](https://beta.reactjs.org/reference/react/useDeferredValue) that is better suited for this purpose, but it would probably require splitting `App` up by creating another component.